pandas/core/tools/datetimes.py

```diff
@@ -60,7 +60,40 @@ def _maybe_cache(arg, format, cache, convert_listlike):
     return cache_array
 
 
-def _convert_and_box_cache(arg, cache_array, box, errors, name=None):
+def _box_if_needed(dt_array, box, default, tz, name=None):
+    """
+    Properly boxes the ndarray of datetimes (if requested) to DatetimeIndex
+    if it is possible or to generic Index instead
+
+    Parameters
+    ----------
+    dt_array: 1-d array
+        array of datetimes to be boxed
+    box : boolean
+        True boxes result as an Index-like, False returns an ndarray
+    tz : object
+        None or 'utc'
+    name : string, default None
+        Name for a resulting index
+
+    Returns
+    -------
+    result : datetime of converted dates
+        Returns:
+
+        - Index-like if box=True
+        - ndarray if box=False
+    """
+    if box:
+        from pandas import DatetimeIndex, Index
+        if is_datetime64_dtype(dt_array):
+            return DatetimeIndex(dt_array, tz=tz, name=name)
+        # e.g. an Index of datetime objects
+        return Index(dt_array, name=name)
+    return default
+
+
+def _convert_and_box_cache(arg, cache_array, box, name=None):
```

Comment:
can you annotate these

Reply:
this might be hard, as `arg` can accept a wide range of types (according to `_convert_and_box_cache` docstring)

Reply:
do what you can, all new code is going to require annotations

Reply:
we have some aggregate annotations in pandas._typing; looking to add some more

Reply:
annotations were added